swim/self_evict.go

```diff
@@ -0,0 +1,285 @@
+	// phase: evicting
+	numberOfPings           int
+	numberOfSuccessfulPings int32
+}
+
+type selfEvict struct {
+	// lock is used to gate access to state changing operations like registering
+	// hooks and starting self eviction. After eviction has started all state
+	// changing functions should return ErrSelfEvictionInProgress
+	lock sync.Mutex
+
+	node    *Node
+	options SelfEvictOptions
+	logger  bark.Logger
+	phases  []*phase
+
+	hooks []SelfEvictHook
+}
+
+func newSelfEvict(node *Node, options SelfEvictOptions) *selfEvict {
+	return &selfEvict{
+		node:    node,
+		options: options,
+		logger:  node.logger,
+	}
+}
+
+// RegisterSelfEvictHook registers a pre/post eviction hook pair. If the hook
+// has already been registered before it returns the ErrDuplicateHook and does
+// not register it twice.
+func (s *selfEvict) RegisterSelfEvictHook(hook SelfEvictHook) error {
+	// lock the phases to prevent reading the current phase in a race. Unlocking
+	// will happen when this function terminates to prevent self-eviction to
+	// start while a hook is being added
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	if s.currentPhase() != nil {
+		return ErrSelfEvictionInProgress
+	}
+
+	for _, h := range s.hooks {
+		if h == hook {
+			return ErrDuplicateHook
+		}
+	}
+
+	s.hooks = append(s.hooks, hook)
+
+	return nil
+}
+
+func (s *selfEvict) SelfEvict() error {
+	s.lock.Lock()
+	if s.currentPhase() != nil {
+		s.lock.Unlock()
+		return ErrSelfEvictionInProgress
+	}
+	// pre-evict
+	s.transitionTo(preEvict)
+	// unlock only after transitioning the phase to make it an atomic operation.
+	s.lock.Unlock()
+
+	s.logger.Info("ringpop is initiating self eviction sequence")
+	s.runHooks(SelfEvictHook.PreEvict)
+
+	s.evict()
+
+	// post-evict
+	s.transitionTo(postEvict)
+	s.runHooks(SelfEvictHook.PostEvict)
+
+	s.done()
+
+	return nil
+}
+
+func (s *selfEvict) evict() {
+	phase := s.transitionTo(evicting)
+	s.node.memberlist.SetLocalStatus(Faulty)
+
+	if s.options.DisablePing == false {
+		numberOfPingableMembers := s.node.memberlist.NumPingableMembers()
+		maxNumberOfPings := int(math.Ceil(float64(numberOfPingableMembers) * s.options.PingRatio))
+
+		// final number of members to ping should not exceed any of:
+		numberOfPings := util.Min(
+			s.node.disseminator.maxP, // the piggyback counter
+			numberOfPingableMembers,  // the number of members we can ping
+			maxNumberOfPings,         // a configured percentage of members
+		)
+
+		phase.numberOfPings = numberOfPings
+
+		// select the members we are going to ping
+		targets := s.node.memberlist.RandomPingableMembers(numberOfPings, nil)
+
+		s.logger.WithFields(bark.Fields{
+			"numberOfPings": phase.numberOfPings,
+			"targets":       targets,
+		}).Debug("starting proactive gossip on self evict")
+
+		var wg sync.WaitGroup
+		wg.Add(len(targets))
+		for _, target := range targets {
+			go func(target Member) {
+				defer wg.Done()
+				_, err := sendPing(s.node, target.address(), s.node.pingTimeout)
```

Comment:
Note: this could result in sending a (other) change more than `maxPiggyBackCounter`. Don't think that it's a problem (it shouldn't) but wanted to point it out to double check.